test/test-prepare-check.c

```diff
@@ -0,0 +1,162 @@
+    assert(check_callback_1_count == 0);
+    assert(check_callback_2_count == 0);
+    assert(timeout_callback_count == 0);
+    break;
+  case 1:
+    /* second iteration: all callbacks should have fired once */
+    assert(prepare_callback_2_count == 1);
+    assert(check_callback_1_count == 1);
+    assert(check_callback_2_count == 1);
+    assert(timeout_callback_count == 1);
+    break;
+  default:
+    /* there should be no third iteration */
+    assert(0);
+  }
+  ++prepare_callback_1_count;
+}
+
+void
+prepare_callback_2(struct event_base *base, struct event_watcher *watcher)
+{
+  printf("prepare_callback_2\n");
+  /* only the first prepare callback should have fired previously, and there should be no second iteration */
+  assert(prepare_callback_1_count == 1);
+  assert(prepare_callback_2_count == 0);
+  assert(check_callback_1_count == 0);
+  assert(check_callback_2_count == 0);
+  assert(timeout_callback_count == 0);
+  ++prepare_callback_2_count;
+}
+
+void
+check_callback_1(struct event_base *base, struct event_watcher *watcher)
+{
+  printf("check_callback_1\n");
+  switch (check_callback_1_count) {
+  case 0:
+    /* first iteration: prepare callbacks should fire before check callbacks */
+    assert(prepare_callback_1_count == 1);
+    assert(prepare_callback_2_count == 1);
+    assert(check_callback_2_count == 0);
+    assert(timeout_callback_count == 0);
+    break;
+  case 1:
+    /* second iteration: first prepare callback should have fired twice */
+    assert(prepare_callback_1_count == 2);
+    assert(prepare_callback_2_count == 1);
+    assert(check_callback_2_count == 1);
+    assert(timeout_callback_count == 1);
+    break;
+  default:
+    /* there should be no third iteration */
+    assert(0);
+  }
+  ++check_callback_1_count;
+}
+
+void
+check_callback_2(struct event_base *base, struct event_watcher *watcher)
+{
+  printf("check_callback_2\n");
+  /* first check callback should fire before second check callback, and there should be no second iteration */
+  assert(prepare_callback_1_count == 1);
+  assert(prepare_callback_2_count == 1);
+  assert(check_callback_1_count == 1);
+  assert(check_callback_2_count == 0);
+  assert(timeout_callback_count == 0);
+  ++check_callback_2_count;
+}
+
+void
+timeout_callback(evutil_socket_t fd, short events, void *arg)
+{
+  printf("timeout_callback\n");
+  /* the prepare and check callbacks should fire before the timeout event, which should only fire once */
+  assert(prepare_callback_1_count == 1);
+  assert(prepare_callback_2_count == 1);
+  assert(check_callback_1_count == 1);
+  assert(check_callback_2_count == 1);
+  assert(timeout_callback_count == 0);
+
+  /* free the second prepare and check watchers to verify they aren't invoked a second time */
+  event_watcher_free(prepare_callback_2_watcher);
+  event_watcher_free(check_callback_2_watcher);
+  ++timeout_callback_count;
+}
+
+int
+main(int argc, char **argv)
+{
+	base = event_base_new();
```

Comment:
Nit: weird indent. In fact, the indent in the the test file is generally shorter than the implementation..